### PR TITLE
Add TimeSpan Properties to BaseResponse for Rate Limit Headers

### PR DESCRIPTION
**Pull request by RealStillkill**
# Abstract

The current headers return strings such as `12m30s`. If you want to consume this data, it needs to be parsed into a more usable format. This PR will add properties to the abstract class BaseResponse that will parse the header strings into something more useful.

# Changes
- Two new public readonly properties of type TimeSpan have been added to BaseResponse, `ResetTokensTimespan ` and `ResetTokensTimespan `
- The original properties have not been changed in order to avoid any breaking changes for downstream projects that may be relying on them.

# Testing Methods
Since regex done poorly can be a massive hindrance in terms of performance, the regex's performance has been tested using a separate project. The regex was run through 100k randomly generated timestamps. The code can be found [here](https://github.com/RealStillkill/OpenAI-DotNet-RegexMetrics). The compute time of the regex was negligible.

This change was based on an extension method used in a private project. It has not produced any problems during its usage in the other project.

### CODE REVIEW

File: OpenAI-DotNet/Common/BaseResponse.cs [hunk 1]
```diff
@@ -68,12 +69,86 @@ public abstract class BaseResponse
...
         /// The time until the rate limit (based on tokens) resets to its initial state.
         /// </summary>
         [JsonIgnore]
         public string ResetTokens { get; internal set; }
 
+        /// <summary>
+        /// The time until the rate limit (based on tokens) resets to its initial state represented as a TimeSpan.
+        /// </summary>
+        [JsonIgnore]
+        public TimeSpan ResetTokensTimespan { get => ConvertTimestampToTimespan(ResetTokens); }
```
Comment (StephenHodgson) [inline OpenAI-DotNet/Common/BaseResponse.cs:88]: can probably be even simpler:
```suggestion
        public TimeSpan ResetTokensTimespan => ConvertTimestampToTimespan(ResetTokens);
```

File: OpenAI-DotNet/Common/BaseResponse.cs [hunk 2]
```diff
@@ -68,12 +69,86 @@ public abstract class BaseResponse
...
+             *  
+             *  Regex Performance Notes, against 100k randomly generated timestamps:
+             *  Average performance: 0.0003ms
+             *  Best case: 0ms
+             *  Worst Case: 15ms
+             *  Total Time: 30ms
+             *  
+             *  Inconsequential compute time
+             */
+            Regex tsRegex = new Regex(@"^(?<hour>\d+h)?(?<mins>\d+m(?!s))?(?<secs>\d+s)?(?<ms>\d+ms)?");
```
Comment (StephenHodgson) [inline OpenAI-DotNet/Common/BaseResponse.cs:113]: this can probably be a private readonly field

File: OpenAI-DotNet/Common/BaseResponse.cs (hunk 1 above)
Comment (RealStillkill) [reply OpenAI-DotNet/Common/BaseResponse.cs:88]: Interesting. I did not know that this syntax exists. I'll commit these changes in a moment.

Comment (RealStillkill) [reply OpenAI-DotNet/Common/BaseResponse.cs]: Change committed

File: OpenAI-DotNet/Common/BaseResponse.cs (hunk 2 above)
Comment (RealStillkill) [reply OpenAI-DotNet/Common/BaseResponse.cs]: Change committed

**Comment by StephenHodgson** [comment]
Thanks I'll add these changes to the next release.

File: OpenAI-DotNet/Common/BaseResponse.cs [hunk 3]
```diff
@@ -3,11 +3,13 @@
 using OpenAI.Extensions;
 using System;
 using System.Text.Json.Serialization;
+using System.Text.RegularExpressions;
 
 namespace OpenAI
 {
     public abstract class BaseResponse
     {
+
```
Comment (StephenHodgson) [inline OpenAI-DotNet/Common/BaseResponse.cs:12]:
```suggestion
```

File: OpenAI-DotNet/Common/BaseResponse.cs [hunk 4]
```diff
@@ -68,12 +70,90 @@ public abstract class BaseResponse
...
         [JsonIgnore]
         public string ResetTokens { get; internal set; }
 
+        /// <summary>
+        /// The time until the rate limit (based on tokens) resets to its initial state represented as a TimeSpan.
+        /// </summary>
+        [JsonIgnore]
+        public TimeSpan ResetTokensTimespan => ConvertTimestampToTimespan(ResetTokens);
+
+
```
Comment (StephenHodgson) [inline OpenAI-DotNet/Common/BaseResponse.cs:91]:
```suggestion
```

File: OpenAI-DotNet/Common/BaseResponse.cs [hunk 5]
```diff
@@ -68,12 +69,88 @@ public abstract class BaseResponse
...
+        *  Regex Performance Notes, against 100k randomly generated timestamps:
+        *  Average performance: 0.0003ms
+        *  Best case: 0ms
+        *  Worst Case: 15ms
+        *  Total Time: 30ms
+        *  
+        *  Inconsequential compute time
+        */
+        private readonly Regex timestampRegex = new Regex(@"^(?<hour>\d+h)?(?<mins>\d+m(?!s))?(?<secs>\d+s)?(?<ms>\d+ms)?");
+
```
Comment (StephenHodgson) [inline OpenAI-DotNet/Common/BaseResponse.cs:106]:
```suggestion
```

File: OpenAI-DotNet/Common/BaseResponse.cs [hunk 6]
```diff
@@ -68,12 +69,87 @@ public abstract class BaseResponse
...
+        /// <summary>
+        /// Takes a timestamp recieved from a OpenAI response header and converts to a TimeSpan
+        /// </summary>
+        /// <param name="timestamp">The timestamp received from an OpenAI header, eg x-ratelimit-reset-tokens</param>
+        /// <returns>A TimeSpan that represents the timestamp provided</returns>
+        /// <exception cref="ArgumentException">Thrown if the provided timestamp is not in the expected format, or if the match is not successful.</exception>
+        private TimeSpan ConvertTimestampToTimespan(string timestamp)
+        {
+            Match match = timestampRegex.Match(timestamp);
+            if (!match.Success)
```
Comment (StephenHodgson) [inline OpenAI-DotNet/Common/BaseResponse.cs:116]:
```suggestion
            Match match = timestampRegex.Match(timestamp);

            if (!match.Success)
```

File: OpenAI-DotNet/Common/BaseResponse.cs [hunk 7]
```diff
@@ -68,12 +69,87 @@ public abstract class BaseResponse
...
+            }
+
+            /*
+             * Note about Hours in timestamps:
+             *  I have not personally observed a timestamp with an hours segment (eg. 1h30m15s1ms).
+             *  Although their presense may not actually exist, we can still have this section in the parser, there is no
+             *  negative impact for a missing hours segment because the capture groups are flagged as optional.
+             */
+            int hours = 0;
+            if (match.Groups.ContainsKey("hour"))
```
Comment (StephenHodgson) [inline OpenAI-DotNet/Common/BaseResponse.cs:128]:
```suggestion
            int hours = 0;

            if (match.Groups.ContainsKey("hour"))
```

File: OpenAI-DotNet/Common/BaseResponse.cs [hunk 8]
```diff
@@ -68,12 +69,87 @@ public abstract class BaseResponse
...
+            {
+                hours = int.Parse(match.Groups["hour"].Value.Replace("h", string.Empty));
+            }
+
+            int minutes = 0;
+            if (match.Groups.ContainsKey("mins"))
```
Comment (StephenHodgson) [inline OpenAI-DotNet/Common/BaseResponse.cs:134]:
```suggestion
            int minutes = 0;

            if (match.Groups.ContainsKey("mins"))
```

File: OpenAI-DotNet/Common/BaseResponse.cs [hunk 9]
```diff
@@ -68,12 +69,87 @@ public abstract class BaseResponse
...
+            {
+                minutes = int.Parse(match.Groups["mins"].Value.Replace("m", string.Empty));
+            }
+
+            int seconds = 0;
+            if (match.Groups.ContainsKey("secs"))
```
Comment (StephenHodgson) [inline OpenAI-DotNet/Common/BaseResponse.cs:140]:
```suggestion
            int seconds = 0;

            if (match.Groups.ContainsKey("secs"))
```

File: OpenAI-DotNet/Common/BaseResponse.cs [hunk 10]
```diff
@@ -68,12 +69,87 @@ public abstract class BaseResponse
...
+            {
+                seconds = int.Parse(match.Groups["secs"].Value.Replace("s", string.Empty));
+            }
+            int ms = 0;
+            if (match.Groups.ContainsKey("ms"))
```
Comment (StephenHodgson) [inline OpenAI-DotNet/Common/BaseResponse.cs:145]:
```suggestion
            int ms = 0;

            if (match.Groups.ContainsKey("ms"))
```

**Pull request merged**